torchrl/envs/libs/openspiel.py

```diff
@@ -0,0 +1,358 @@
+        >>> print(env.available_envs)
+        ['2048', 'add_noise', 'amazons', 'backgammon', ...]
+    """
+
+    git_url = "https://github.com/google-deepmind/open_spiel"
+    libname = "pyspiel"
+    _lib = None
+
+    @_classproperty
+    def lib(cls):
+        if cls._lib is not None:
+            return cls._lib
+
+        import pyspiel
+
+        cls._lib = pyspiel
+        return pyspiel
+
+    @_classproperty
+    def available_envs(cls):
+        if not _has_pyspiel:
+            return []
+        return _get_envs()
+
+    def __init__(self, env=None, categorical_action_encoding=True, **kwargs):
+        if env is not None:
+            kwargs["env"] = env
+        self._seed_calls_reset = None
+        if not categorical_action_encoding:
+            raise NotImplementedError
+        self._categorical_action_encoding = categorical_action_encoding
+        super().__init__(**kwargs)
+
+    def _check_kwargs(self, kwargs: Dict):
+        pyspiel = self.lib
+        if "env" not in kwargs:
+            raise TypeError("Could not find environment key 'env' in kwargs.")
+        env = kwargs["env"]
+        if not isinstance(env, pyspiel.State):
+            raise TypeError("env is not of type 'pyspiel.State'.")
+
+    def _build_env(self, env, requires_grad: bool = False, **kwargs):
+        self.requires_grad = requires_grad
+        return env
+
+    def _init_env(self):
+        self._update_action_mask()
+
+    def _make_specs(self, env: "pyspiel.State") -> None:  # noqa: F821
+        self.reward_spec = UnboundedContinuousTensorSpec(
+            shape=[*self.batch_size, env.num_players()],
+            device=self.device,
+        )
+        self.done_spec = DiscreteTensorSpec(
+            n=2,
+            shape=torch.Size((1,)),
+            dtype=torch.bool,
+            device=self.device,
+        )
+        action_spec_cls = (
+            DiscreteTensorSpec
+            if self._categorical_action_encoding
+            else OneHotDiscreteTensorSpec
+        )
+        self.action_spec = action_spec_cls(
+            env.num_distinct_actions(),
+            dtype=torch.int64,
+            device=self.device,
+        )
+        self.observation_spec = CompositeSpec(
+            observation=UnboundedContinuousTensorSpec(
+                shape=(
+                    *self.batch_size,
+                    env.num_players(),
+                    *env.get_game().observation_tensor_shape(),
+                ),
+                device=self.device,
+            ),
+            current_player=DiscreteTensorSpec(
+                env.num_players(),
+                dtype=torch.int,
+                device=self.device,
+            ),
+            state=NonTensorSpec([]),
+            shape=self.batch_size,
+        )
+
+    def _set_seed(self, seed):
+        if seed is not None:
+            raise NotImplementedError("This environment has no seed.")
+
+    def current_player(self):
+        return self._env.current_player()
```

Comment:
@matteobettini if we have multiple players, do we have to consider that as a MARL problem?
What's the proper way to represent the env here?

I see various options:
- what we are currently doing, tracking the player in the TD
- Having a nested TD with `"player0"` and `"player1"` (should that be dynamic)
- i also thought: can we have 2 envs that play against each other? I guess for MCTS that could be nice to keep things separated

Reply:
In pettingzoo we have some boatd games already. I guess the choice here also depends on if the game is turn-based or not (players can  act in parallel)

the way we did it in pettingzoo is to keep the players in 2 nested td groups (your option 2) and have a mask for the non-playing one.

Reply:
the goal is to use it for MCTS so I'd say turn based.
Self-play sort of thing

Reply:
I will push an update to put the players in their own nested td groups with a mask, like in PettingZooEnv